File: Java/src/com/leetcode/Medium/RemoveNthNodeFromEndOfList.java
package com.leetcode.Medium;

// 19. Remove Nth Node From End of List
public class RemoveNthNodeFromEndOfList {

    static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null || n < 1) {
            return head;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy;
        ListNode slow = dummy;
        while (fast != null && n + 1 > 0) {
            fast = fast.next;
            n--;
        }
        while (fast != null) {
            slow = slow.next;
            fast = fast.next;
        }
        slow.next = slow.next.next;
        return dummy.next;
    }
}

File: JavaScript/Medium/RemoveNthNodeFromEndOfList.js
/**
 * 19. Remove Nth Node From End of List
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  if (n == 0) return head;

  let len = getLength(head);
  if (n >= len) return head.next;

  let curr = head;
  let i = 0, j = len - n - 1;
  while (i < j) {
    i++;
    curr = curr.next;
  }

  curr.next = curr.next.next;
  return head;
};

const getLength = function (node) {
  let len = 0;
  while (node != null) {
    len++;
    node = node.next;
  }
  return len;
};
